src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
             {
-                get
-                {
-                    return _f;
-                }
+                Debug.Assert(float.IsFinite(value));
+                Debug.Assert(value > 0.0f);
+                f = ExtractFractionAndBiasedExponent(value, out e);
+            }
 
-                set
-                {
-                    _f = value;
-                }
+            // Computes the two boundaries of value.
+            //
+            // The bigger boundary (mPlus) is normalized.
+            // The lower boundary has the same exponent as mPlus.
+            //
+            // Precondition:
+            //  The value encoded by value must be greater than 0.
+            public DiyFp(double value, out DiyFp mMinus, out DiyFp mPlus) : this(value)
+            {
+                GetBoundaries(52, out mMinus, out mPlus);
             }
 
-            public int e
+            // Computes the two boundaries of value.
+            //
+            // The bigger boundary (mPlus) is normalized.
+            // The lower boundary has the same exponent as mPlus.
+            //
+            // Precondition:
+            //  The value encoded by value must be greater than 0.
+            public DiyFp(float value, out DiyFp mMinus, out DiyFp mPlus) : this(value)
             {
-                get
-                {
-                    return _e;
-                }
+                GetBoundaries(23, out mMinus, out mPlus);
+            }
 
-                set
-                {
-                    _e = value;
-                }
+            public DiyFp(ulong significand, int exponent)
+            {
+                f = significand;
+                e = exponent;
             }
 
-            public static void GenerateNormalizedDiyFp(double value, out DiyFp result)
+            public DiyFp Multiply(in DiyFp other)
             {
-                Debug.Assert(value > 0.0);
+                // Simply "emulates" a 128-bit multiplcation
+                //
+                // However: the resulting number only contains 64-bits. The least
+                // signficant 64-bits are only used for rounding the most significant
+                // 64-bits.
 
-                long f = ExtractFractionAndBiasedExponent(value, out int e);
+                uint a = (uint)(f >> 32);
+                uint b = (uint)(f);
 
-                while ((f & (1L << 52)) == 0)
-                {
-                    f <<= 1;
-                    e--;
-                }
+                uint c = (uint)(other.f >> 32);
+                uint d = (uint)(other.f);
+
+                ulong ac = ((ulong)(a) * c);
+                ulong bc = ((ulong)(b) * c);
+                ulong ad = ((ulong)(a) * d);
+                ulong bd = ((ulong)(b) * d);
 
-                int lengthDelta = (SignificandLength - 53);
-                f <<= lengthDelta;
-                e -= lengthDelta;
+                ulong tmp = (bd >> 32) + (uint)(ad) + (uint)(bc);
```

Comment:
What's `tmp` here? Can we represent this with a better name esp since we seem to be modifying it later?

Reply:
This, as the function describes above, is doing a 128-bit multiplication but only returning the upper 64-bits with the appropriate rounding.
There isn't really a meaningful name for `tmp` as it is doing multiple steps at once, on the appropriate bits, so we can round the upper-64bits accordingly.

`tmp` is also the name used in the reference sources.